### DISCUSSION TITLE
How to use probe sockops or sk_msg in bpftrace

### DISCUSSION THREAD

**Discussion by zzjianhui** [bpftrace/bpftrace#2649] [Q&A]
No sockops is displayed when I use the bpftrace -l command.
When I use sockops in bpftrace, "ERROR: Unrecognized probe type: sockops" is displayed.
So what type of probe does sockops belong to? and how to use?

**Comment by viktormalik** [comment]
bpftrace does not support the socket ops program type at this moment.

**Reply by zzjianhui** [reply]
Is BCC supported? Or only C eBPF support?

**Reply by danobi** [reply]
To clarify, sockopts are not really a tracing program type -- it's more used for TCP tuning IIUC. So bpftrace does not support it. BCC may support it; you would need to check. And yeah, C would probably be a safe choice to write a sockops program. I'm not sure the rust support is quite there for it yet.

**Comment by chengshuyi** [comment]
you can use libbpf-rs. i have some demo about sockops and sk_msg. see [sk_msg](https://github.com/chengshuyi/bpfsamples/tree/master/sk_msg) and [sockops](https://github.com/chengshuyi/bpfsamples/tree/master/sockops)

**Reply by zzjianhui** [reply]
Only C eBPF can support?

**Reply by chengshuyi** [reply]
Yes.
